arrows/vxl/aligned_edge_detection.cxx

```diff
@@ -0,0 +1,304 @@
+    vil_plane( all_channels, 1 ) = vil_plane( aligned_edges, 1 );
+    vil_plane( all_channels, 2 ) = joint_nms_edges;
+    return all_channels;
+  }
+  return aligned_edges;
+}
+
+// ----------------------------------------------------------------------------
+aligned_edge_detection
+::aligned_edge_detection()
+  : d( new priv( this ) )
+{
+  attach_logger( "arrows.vxl.aligned_edge_detection" );
+}
+
+aligned_edge_detection
+::~aligned_edge_detection()
+{
+}
+
+vital::config_block_sptr
+aligned_edge_detection
+::get_configuration() const
+{
+  // get base config from base class
+  vital::config_block_sptr config = algorithm::get_configuration();
+
+  config->set_value( "threshold",
+                     d->threshold,
+                     "Minimum edge magnitude required to report as an edge "
+                     "in any output image." );
+  config->set_value( "produce_joint_output",
+                     d->produce_joint_output,
+                     "Set to false if we do not want to spend time computing "
+                     "joint edge images comprised of both horizontal and vertical "
+                     "information." );
+  config->set_value( "smoothing_sigma",
+                     d->smoothing_sigma,
+                     "Smoothing sigma for the output NMS edge density map." );
+  config->set_value( "smoothing_half_step",
+                     d->smoothing_half_step,
+                     "Smoothing half step for the output NMS edge density map." );
+
+  return config;
+}
+
+// ----------------------------------------------------------------------------
+void
+aligned_edge_detection
+::set_configuration( vital::config_block_sptr in_config )
+{
+  // Starting with our generated vital::config_block to ensure that assumed
+  // values are present. An alternative is to check for key presence before
+  // performing a get_value() call.
+  vital::config_block_sptr config = this->get_configuration();
+  config->merge_config( in_config );
+
+  // Settings for edge detection
+  d->threshold = config->get_value< float >( "threshold" );
+  d->produce_joint_output =
+    config->get_value< bool >( "produce_joint_output" );
+  d->smoothing_sigma = config->get_value< double >( "smoothing_sigma" );
+  d->smoothing_half_step =
+    config->get_value< unsigned >( "smoothing_half_step" );
+}
+
+// ----------------------------------------------------------------------------
+bool
+aligned_edge_detection
+::check_configuration( vital::config_block_sptr config ) const
+{
+  return true;
+}
+
+// ----------------------------------------------------------------------------
+kwiver::vital::image_container_sptr
+aligned_edge_detection
+::filter( kwiver::vital::image_container_sptr image_data )
+{
+  // Get input image
+  vil_image_view_base_sptr source_image =
+    vxl::image_container::vital_to_vxl( image_data->get_image() );
+
+  // Perform Basic Validation
+  if( !image_data )
+  {
+    return kwiver::vital::image_container_sptr();
```

Comment:
```suggestion
    return {};
```

...or possibly `return nullptr;`. (Multiple instances.)

Reply:
I'm open to suggestions between the two. I'd hope a clear error message mitigates any confusion, so I'm not sure it matters much.

Reply:
There is no effective difference; all return a null shared pointer (and at least at -O2 or higher should produce the same assembly). It's just a question of whether it's worthwhile to spell out the full type and whether `nullptr` is clearer than a default-constructed `shared_ptr`.

Reply:
I changed everything to `nullptr`.